### PR TITLE
Update signup error messages

### LINKED ISSUE #10460: Error message during sign-up for too long/too short usernames is uninformative
Error message during sign-up is uninformative -- instead of showing that the length of the username is insufficient/too long, it just says "Invalid request parameters."

image

### PR DESCRIPTION
**Pull request by thefifthisa**
[//]: # (Note: See http://habitica.wikia.com/wiki/Using_Your_Local_Install_to_Modify_Habitica%27s_Website_and_API for more info)

[//]: # (Put Issue # here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes #10460 

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)

Updated the error messages to say exactly what's wrong with the signup information provided (instead of just `Invalid request parameters.`).

~~However, this fix doesn't work when the password is missing.~~

[//]: # (Put User ID in here - found on the Habitica website at User Icon > Settings > API)

----
UUID: 9ca2bb65-ce6c-4e4c-a367-5ef9b5c7421e

### CODE REVIEW

File: test/api/unit/middlewares/errorHandler.test.js [hunk 1]
```diff
@@ -123,7 +123,7 @@ describe('errorHandler', () => {
     expect(res.json).to.be.calledWith({
       success: false,
       error: 'BadRequest',
-      message: 'Invalid request parameters.',
+      message: 'invalid param',
```
Comment (thefifthisa) [inline test/api/unit/middlewares/errorHandler.test.js]: Update unit test to expect the actual invalid parameter instead of the generic message.

File: website/server/middlewares/errorHandler.js [hunk 2]
```diff
@@ -83,6 +83,15 @@ module.exports = function errorHandler (err, req, res, next) { // eslint-disable
 
   if (responseErr.errors) {
     jsonRes.errors = responseErr.errors;
+    if (jsonRes.message === res.t('invalidReqParams')) {
```
Comment (thefifthisa) [inline website/server/middlewares/errorHandler.js]: Target only `invalidReqParams` errors when reformatting the error messages.

File: website/server/controllers/api-v3/auth.js [hunk 3]
```diff
@@ -101,23 +101,48 @@ api.registerLocal = {
...
       },
+      email: {
+        notEmpty: true,
+        errorMessage: res.t('missingEmail'),
+        isEmail: {errorMessage: res.t('notAnEmail')},
+      },
       password: {
-        notEmpty: {errorMessage: res.t('missingPassword')},
+        notEmpty: true,
+        errorMessage: res.t('missingPassword'),
```
Comment (thefifthisa) [inline website/server/controllers/api-v3/auth.js]: Setting `notEmpty` to true makes the `missingUsername`, `missingEmail`, and `missingPassword` errors show up when these fields are empty.

File: website/server/middlewares/errorHandler.js (hunk 2 above)
Comment (paglias) [inline website/server/middlewares/errorHandler.js]: So there's a problem with this change, it shouldn't change the actual message returned, the API already returns the specific error messages in the `errors` property of `jsonRes`, you should use its content to show a different message on the client side instead of changing the server side message

File: website/server/controllers/api-v3/auth.js [hunk 4]
```diff
@@ -101,23 +101,48 @@ api.registerLocal = {
...
+          errors.push(e);
+          emailCheck = true;
+        }
+        if (!passwordCheck && e.param === 'password') {
+          errors.push(e);
+          passwordCheck = true;
+        }
+      }
+      throw errors;
+    }
```
Comment (thefifthisa) [inline website/server/controllers/api-v3/auth.js]: Not sure if this is the most efficient way to do this, so any suggestions are welcome!

File: website/client/components/auth/registerLoginReset.vue [hunk 5]
```diff
@@ -402,11 +402,6 @@ export default {
       window.location.href = redirectTo;
     },
     async login () {
-      if (!this.username) {
-        alert('Email is required');
-        return;
-      }
-
```
Comment (thefifthisa) [inline website/client/components/auth/registerLoginReset.vue]: The alert is now a snackbar notification like other validation errors.

File: website/client/components/static/home.vue [hunk 6]
```diff
@@ -616,11 +616,6 @@
       },
       // @TODO this is totally duplicate from the registerLogin component
       async register () {
-        if (this.password !== this.passwordConfirm) {
-          alert('Passwords must match');
-          return;
-        }
-
```
Comment (thefifthisa) [inline website/client/components/static/home.vue]: The alert is now a snackbar notification like other validation errors.

**Comment by thefifthisa** [comment]
I think I got it! Here are some gifs of how an invalid [username](https://user-images.githubusercontent.com/30595954/42124337-4adf9622-7c93-11e8-9bff-c99383241315.gif), [email](https://user-images.githubusercontent.com/30595954/42124338-4eb17f86-7c93-11e8-9bad-8fcd58ff688b.gif), or [password](https://user-images.githubusercontent.com/30595954/42124339-518f9b48-7c93-11e8-9019-95cca66103c7.gif) is handled. Also, multiple errors look like this:

signuperrors

File: website/server/controllers/api-v3/auth.js [hunk 7]
```diff
@@ -101,21 +101,25 @@ api.registerLocal = {
     let existingUser = res.locals.user; // If adding local auth to social user
 
     req.checkBody({
-      email: {
-        notEmpty: {errorMessage: res.t('missingEmail')},
-        isEmail: {errorMessage: res.t('notAnEmail')},
-      },
       username: {
-        notEmpty: {errorMessage: res.t('missingUsername')},
```
Comment (paglias) [inline website/server/controllers/api-v3/auth.js:109]: why was this line changed?

Comment (thefifthisa) [reply website/server/controllers/api-v3/auth.js:109]: Before, when a field was left empty, the error was something like "Invalid email address" or "Login Name must be 1 to 20 characters." Now, with `notEmpty` set to true, it says "Missing email" or "Missing Login Name" as intended.

**Comment by SabreCat** [comment]
Looks great, @thefifthisa! Welcome to Blacksmith tier 2!

**Comment by SabreCat** [comment]
Had to fix one issue where this PR was causing other error messages (e.g. Not enough Gold to buy Armoire, wrong password at login) not to display. https://github.com/HabitRPG/habitica/commit/c630486fef70ce74379663d01525b0b7006eead5 Staging testers, keep an eye out!

**Comment by thefifthisa** [comment]
@SabreCat ahhh, sorry about that!
